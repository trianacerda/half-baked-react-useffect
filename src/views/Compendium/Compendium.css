.app {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: #f6cf57;
}

.title {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 20px;
}

main {
  padding: 20px 10px;
}

.titleText {
  margin-left: 10px;
}
