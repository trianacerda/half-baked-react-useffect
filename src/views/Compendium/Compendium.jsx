import {useState, useEffect} from 'react';
import PokemonList from '../../components/PokemonList/PokemonList';
import {
  fetchFilteredPokemon,
  fetchPokemon,
  fetchSearchPokemon,
  fetchTypes,
} from '../../services/pokemon';
import './Compendium.css';
import Controls from '../../components/Controls/Controls';
import pokeball from '../../assets/pokeball.png';

export default function Compendium() {
  const [loading, setLoading] = useState(true);
  const [pokemons, setPokemons] = useState([]);
  const [searchName, setSearchName] = useState('');
  const [types, setTypes] = useState([]);
  const [selectedType, setSelectedType] = useState('all');

  //TODO ðŸ˜– help!
  // if (pokemons.count !== 0) {
  //   useEffect(() => {
  //     const getPokemon = async () => {
  //       const pokemonList = await fetchPokemon();
  //        this.setState({pokemons: pokemonList});
  //       setLoading(false);
  //     };
  //     getPokemon();
  //   }, [pokemons]);
  // }

  //TODO ðŸ˜– help!
  //   useEffect(async () => {
  //      function getTypes() {
  //       const pokemonTypes = fetchTypes();
  //       setTypes(pokemonTypes);
  //     }
  //     getTypes();
  //   }, []);

  //TODO ðŸ˜– help!
  // useEffect(() => {
  //   async function getFilteredPokemon() {
  //     if (!selectedType) return;
  //     setLoading(true);

  //     if (selectedType !== 'all') {
  //       const filteredPokemon = await fetchFilteredPokemon(selectedType);
  //       setPokemons(filteredPokemon);
  //     } else {
  //       const pokemonList = await fetchPokemon();
  //       this.setState({pokemons: pokemonList});
  //     }
  //     setLoading(false);
  //     setSort('');
  //   }

  //   getFilteredPokemon();
  // }, [selectedType]);

  const handleSubmit = (event) => {
    event.preventDefault();
    setLoading(true);
    fetchSearchPokemon(searchName)
      .then((searchedPokemons) => {
        this.setState({pokemons: searchedPokemons});
      })
      .catch((error) => {})
      .finally(() => {
        setLoading(false);
        setSearchName('');
        setSelectedType('');
      });
  };

  return (
    <div className='app'>
      <main>
        <div className='title'>
          <img src={pokeball} alt='pokeball' />
          <h1 className='titleText'>Alchemy Compendium</h1>
        </div>
        <Controls
          name={searchName}
          handleSubmit={handleSubmit}
          handleNameChange={setSearchName}
          types={types}
          filterChange={setSelectedType}
          selectedType={selectedType}
        />
        {loading ? (
          <code>Search for the bugs in the code!</code>
        ) : (
          <PokemonList pokemons={pokemons} />
        )}
      </main>
    </div>
  );
}
