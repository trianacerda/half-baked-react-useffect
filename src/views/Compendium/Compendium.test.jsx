import { screen, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Compendium from './Compendium'

it('Should fetch then render a list of pokemon', async () => {
  render(<Compendium />)

  const loading = screen.getByText(/Loading.../i)
  expect(loading).toBeInTheDocument()

  const pokemons = await screen.findByRole('list', {
    name: 'pokemonList',
  })

  expect(pokemons).toMatchSnapshot()
})

it('should display a search form', async () => {
  render(<Compendium />)
  const searchForm = await screen.findByRole('form', {
    name: 'searchForm',
  })
  expect(searchForm).toBeInTheDocument()
})

it('should search for a pokemon', async () => {
  render(<Compendium />)

  const searchInput = await screen.findByLabelText('Name:')
  userEvent.type(searchInput, 'Charizard')

  const submitButton = await screen.findByRole('button')
  userEvent.click(submitButton)

  return waitFor(() => {
    const pokemons = screen.getAllByText('Charizard', { exact: false })

    expect(pokemons).toHaveLength(3)
    expect(pokemons).toMatchSnapshot()
  })
})
