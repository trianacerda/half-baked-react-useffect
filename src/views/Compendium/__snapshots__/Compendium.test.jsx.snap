// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should fetch then render a list of pokemon 1`] = `
<ul
  aria-label="pokemonList"
  class="pokemonList"
>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Charizard-First Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/006.png"
      />
      <div
        class="details"
      >
        <h2>
          Charizard
        </h2>
        <p>
          First Generation
        </p>
        <p>
          Blaze
        </p>
        <p>
          Fire
          /Flying
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Typhlosion-Second Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/157.png"
      />
      <div
        class="details"
      >
        <h2>
          Typhlosion
        </h2>
        <p>
          Second Generation
        </p>
        <p>
          Blaze
        </p>
        <p>
          Fire
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Blaziken-Third Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/257.png"
      />
      <div
        class="details"
      >
        <h2>
          Blaziken
        </h2>
        <p>
          Third Generation
        </p>
        <p>
          Blaze
        </p>
        <p>
          Fire
          /Fighting
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Mewtwo-First Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/150.png"
      />
      <div
        class="details"
      >
        <h2>
          Mewtwo
        </h2>
        <p>
          First Generation
        </p>
        <p>
          Pressure
        </p>
        <p>
          Psychic
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Suicune-Second Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/245.png"
      />
      <div
        class="details"
      >
        <h2>
          Suicune
        </h2>
        <p>
          Second Generation
        </p>
        <p>
          Pressure
        </p>
        <p>
          Water
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Pidgeot-First Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/018.png"
      />
      <div
        class="details"
      >
        <h2>
          Pidgeot
        </h2>
        <p>
          First Generation
        </p>
        <p>
          Keen-eye
          /Tangled-feet
        </p>
        <p>
          Normal
          /Flying
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Golduck-First Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/055.png"
      />
      <div
        class="details"
      >
        <h2>
          Golduck
        </h2>
        <p>
          First Generation
        </p>
        <p>
          Damp
          /Cloud-nine
        </p>
        <p>
          Water
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Gardevoir-Third Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/282.png"
      />
      <div
        class="details"
      >
        <h2>
          Gardevoir
        </h2>
        <p>
          Third Generation
        </p>
        <p>
          Synchronize
          /Trace
        </p>
        <p>
          Psychic
          /Fairy
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Ninetales-First Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/038.png"
      />
      <div
        class="details"
      >
        <h2>
          Ninetales
        </h2>
        <p>
          First Generation
        </p>
        <p>
          Flash-fire
        </p>
        <p>
          Fire
        </p>
      </div>
    </figure>
  </li>
  <li
    class="listItem"
  >
    <figure
      aria-label="pokemon"
    >
      <img
        alt="Houndoom-Second Generation"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/229.png"
      />
      <div
        class="details"
      >
        <h2>
          Houndoom
        </h2>
        <p>
          Second Generation
        </p>
        <p>
          Early-bird
          /Flash-fire
        </p>
        <p>
          Dark
          /Fire
        </p>
      </div>
    </figure>
  </li>
</ul>
`;

exports[`should search for a pokemon 1`] = `
Array [
  <h2>
    Charizard
  </h2>,
  <h2>
    Charizard-mega-x
  </h2>,
  <h2>
    Charizard-mega-y
  </h2>,
]
`;
